### ISSUE TITLE
[SD-XL] Fine-tuning text-to-image script for full U-net

### ISSUE THREAD

**Issue by bghira** [huggingface/diffusers#4366]
**Is your feature request related to a problem? Please describe.**

The current options available for fine-tuning SDXL are currently inadequate for training a new noise schedule into the base U-net.

**Describe the solution you'd like**

I would like a replica of the Stable Diffusion 1.5/2.1 text-to-image scripts, in the style of SDXL's requirements.

**Describe alternatives you've considered**

I have considered a LoRA merge, but the results have not been viable.

**Additional context**

Our goal is to fine-tune the SDXL 1.0 base model on v-prediction as a part of a multi-stage effort to resolve its contrast issues and to make it easier to introduce inpainting models, through zero terminal SNR fine-tuning efforts on large training hardware.

cc @sayakpaul

**Comment by sayakpaul** [comment]
I think the hardware requirements of doing this could be quite restrictive to the community, given that SDXL already takes a lot of resources. This is why, we restricted ourselves to LoRA DreamBooth for the most part. However, we also have ControlNet and InstructPix2Pix trainers. 

But I do see the value proposed here. Happy to review PRs here :)

**Comment by FeiiYin** [comment]
Hope to see the PRs soon! Looking forward to finetuning the unet of XL. :-)

**Comment by AmericanPresidentJimmyCarter** [comment]
> I think the hardware requirements of doing this could be quite restrictive to the community, given that SDXL already takes a lot of resources. 

Kohya trainer works fine on a 3090 and you can train batch sizes of 4x4x128x128 with grad checkpointing on a single card, I don't see why it would be inaccessible to consumers. You might want to a add a script that lets them precompute CLIP embeddings for the dataset, so they don't have to worry about that.

**Comment by bghira** [comment]
@sayakpaul please see comments above. I agree. tuning at 1024px isn't the only use case of fine-tuning u-net.

we also would like to investigate fine-tuning on spectrograms for Riffusion XL. this is something such a script would be useful for.

**Comment by sayakpaul** [comment]
> Kohya trainer works fine on a 3090 and you can train batch sizes of 4x4x128x128 with grad checkpointing on a single card, I don't see why it would be inaccessible to consumers. 

For 128x128 resolution, sure, shouldn't be a problem. 

Sure!

As mentioned, happy to review a PR for this. We could potentially cache the VAE encodings too to save further memory.

**Comment by bghira** [comment]
well, the instruct pix2pix script, uses more than 40GiB of VRAM at 128x128 and a batch size of one.

how would a fine-tune text2img script be any worse?

as much as I would love to submit the script for this, I don't think I can pull that off.

**Comment by sayakpaul** [comment]
Yeah it does but it also serves a different purpose. 

For pure SDXL fine-tuning, we would like to gauge the community interest a bit more.

**Comment by bghira** [comment]
to compare, the instruct pix2pix defaults out of the box call for 256x256, and use:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.116.04   Driver Version: 525.116.04   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100 80G...  Off  | 00000000:00:05.0 Off |                    0 |
| N/A   60C    P0   307W / 300W |  56977MiB / 81920MiB |     96%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

about 56GiB of VRAM...

**Comment by bghira** [comment]
instruct pix2pix at 1024px resolution, batch size 4:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.116.04   Driver Version: 525.116.04   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100 80G...  Off  | 00000000:00:05.0 Off |                    0 |
| N/A   72C    P0   314W / 300W |  69619MiB / 81920MiB |    100%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
```

so if you have the initial bulk of the VRAM required to do 256x256, you can do 1024x1024. it's not that much more VRAM to use - this model is very inefficient in any training configuration.

**Comment by sayakpaul** [comment]
Thanks for providing the information. I am opening it up for the community to contribute but we observe enough interest, we will do it ourselves :)

> it's not that much more VRAM to use - this model is very inefficient in any training configuration.

Could you elaborate? How can we make it more efficient?

**Comment by bghira** [comment]
> Could you elaborate? How can we make it more efficient?

unlikely to be possible.

**Comment by AmericanPresidentJimmyCarter** [comment]
> Thanks for providing the information. I am opening it up for the community to contribute but we observe enough interest, we will do it ourselves :)
> 
> > it's not that much more VRAM to use - this model is very inefficient in any training configuration.
> 
> Could you elaborate? How can we make it more efficient?

Easily.

Write a script called "precompute_latents_and_conds.py" which creates a new HF dataset from a dataset containing captions and images. The new HF dataset contains pre-computed CLIP embeddings and VAE-encoded images. Now just call `self.unet` instead of `pipeline` and backwards on the loss from that. Train in BF16 with gradient checkpointing.

**Comment by AmericanPresidentJimmyCarter** [comment]
> > Kohya trainer works fine on a 3090 and you can train batch sizes of 4x4x128x128 with grad checkpointing on a single card, I don't see why it would be inaccessible to consumers.
> 
> For 128x128 resolution, sure, shouldn't be a problem.
> 
> Sure!
> 
> As mentioned, happy to review a PR for this. We could potentially cache the VAE encodings too to save further memory.

4x4x128x128 is batch size 4, 1024x1024 pixel images for VAE latents

**Comment by Yangr116** [comment]
Looking forward to finetuning the unet of SDXL!

**Comment by CaptnSeraph** [comment]
i think this would make sense, its not hard to get access to equipment over 24gbvram and certainly a lot of trainers are willing to commit resources towards finetunes, might as well open up the capability to those that wish to do it.

Its not as if it will harm those that cannot, for those that cannot use this due to the vram constraints of consumer level devices there is the dreambooth lora.

but for those with bigger hardware or with resources to rent them, it makes sense to allow this level of finetune to be possible.

Plus a zero terminal SNR model would be amazing.

**Comment by sayakpaul** [comment]
Looks like there's indeed interest! 

@CaptnSeraph would you be willing to put together a script and start a PR?

**Comment by CaptnSeraph** [comment]
id love to be the initiator on this, but i do not know enough about the backend stuff on sdxl training to be the first one to put pen to paper.
